causalpy/experiments/interrupted_time_series.py

```diff
@@ -25,16 +26,279 @@
 from patsy import build_design_matrices, dmatrices
 from sklearn.base import RegressorMixin
 
-from causalpy.custom_exceptions import BadIndexException
+from causalpy.custom_exceptions import BadIndexException, ModelException
+from causalpy.experiments.base import BaseExperiment
 from causalpy.plot_utils import get_hdi_to_df, plot_xY
 from causalpy.pymc_models import PyMCModel
 from causalpy.utils import round_num
 
-from .base import BaseExperiment
-
 LEGEND_FONT_SIZE = 12
 
 
+class TreatmentTimeHandler(ABC):
+    @abstractmethod
+    def data_preprocessing(self, data, treatment_time, model):
+        pass
+
+    @abstractmethod
+    def data_postprocessing(
+        self, model, data, idata, treatment_time, y, X, pre_y, pre_X
+    ):
+        pass
+
+    @abstractmethod
+    def plot_intervention_line(
+        self, ax, handles, labels, datapre, datapost, pre_pred, post_pred
+    ):
+        pass
+
+    @abstractmethod
+    def plot_impact_cumulative(self, ax, datapre, datapost, post_impact_cumulative):
+        pass
+
+    def plot_treated_counterfactual(
+        self, ax, handles, labels, datapre, datapost, pre_pred, post_pred
+    ):
+        """Optional: override if needed"""
+        pass
+
+
+class UnknownTreatmentTimeHandler(TreatmentTimeHandler):
+    """
+    A utility class for managing data preprocessing, postprocessing,
+    and plotting steps for models that infer unknown treatment times.
+
+    This handler prepares input data for the model, extracts relevant
+    outputs after inference, and structures them for further analysis
+    and visualization.
+    """
+
+    def data_preprocessing(self, data, treatment_time, model):
+        """
+        Preprocesses the input data by constraining the model's
+        treatment time inference window.
+        """
+        # Restrict model's treatment time inference to given range
+        model.set_time_range(treatment_time, data)
```

Comment:
is treatment time not unknown in preprocessing stage? Does the input here warrant a better name?